modules/exadsBidAdapter.js

```diff
@@ -0,0 +1,433 @@
+    utils.logInfo('bidResponse->', bidResponse);
+
+    bidResponses.push(bidResponse);
+  } else {
+    utils.logInfo('NO Ad server response ->', serverResponse.body.id);
+  }
+
+  utils.logInfo('interpretResponse -> bidResponses:', bidResponses);
+
+  return bidResponses;
+}
+
+function makeBidRequest(url, data) {
+  const payloadString = JSON.stringify(data);
+
+  return {
+    method: 'POST',
+    url: url,
+    data: payloadString,
+  }
+}
+
+function getUrl(adPartner, bid) {
+  let endpointUrlMapping = {
+    [PARTNERS.RTB_2_4]: bid.params.endpoint + '?idzone=' + bid.params.zoneId + '&fid=' + bid.params.fid
+  };
+
+  return endpointUrlMapping[adPartner] ? endpointUrlMapping[adPartner] : 'defaultEndpoint';
+}
+
+function manageEnvParams() {
+  envParams.domain = window.location.hostname;
+  envParams.page = window.location.protocol + '//' + window.location.host + window.location.pathname;
+  envParams.lang = navigator.language.indexOf('-') > -1
+    ? navigator.language.split('-')[0]
+    : navigator.language;
+  envParams.userAgent = navigator.userAgent;
+
+  if (navigator.appVersion.indexOf('Win') !== -1) {
+    envParams.osName = 'Windows';
+  } else if (navigator.appVersion.indexOf('Mac') !== -1) {
+    envParams.osName = 'MacOS';
+  } else if (navigator.appVersion.indexOf('X11') !== -1) {
+    envParams.osName = 'Unix';
+  } else if (navigator.appVersion.indexOf('Linux') !== -1) {
+    envParams.osName = 'Linux';
+  }
+
+  let browserLanguage = navigator.language || navigator.userLanguage;
+  let acceptLanguage = browserLanguage.replace('_', '-');
+
+  envParams.language = acceptLanguage;
+
+  utils.logInfo('Domain -> ', envParams.domain);
+  utils.logInfo('Page -> ', envParams.page);
+  utils.logInfo('Lang -> ', envParams.lang);
+  utils.logInfo('OS -> ', envParams.osName);
+  utils.logInfo('User Agent -> ', envParams.userAgent);
+  utils.logInfo('Primary Language -> ', envParams.language);
+}
+
+export const imps = new Map();
+
+manageEnvParams();
+
+function handleValidRTB2Dot4(bid) {
+  const bannerInfo = bid.mediaTypes?.banner;
+  const nativeInfo = bid.mediaTypes?.native;
+  const videoInfo = bid.mediaTypes?.video;
+  const isValid = !!(
+    bid.params.endpoint &&
+    bid.params.userIp &&
+    bid.params.hasOwnProperty('userId') &&
+    bid.params.zoneId &&
+    bid.params.partner &&
+    bid.params.fid &&
+    bid.params.siteId &&
+    bid.params.impressionId &&
+    bid.params.country &&
+    bid.params.country.length > 0 &&
+    (bannerInfo || nativeInfo || videoInfo) &&
+    !!(bid.params.bidfloor && bid.params.bidfloorcur) &&
+    (nativeInfo ? bid.params.native &&
+      nativeInfo.ortb.assets &&
+      nativeInfo.ortb.assets.some(asset => !!asset.img) : true) &&
+    (videoInfo ? !!(bid.params.stream &&
+      bid.params.stream.video &&
+      bid.params.stream.video.mimes &&
+      bid.params.stream.video.mimes.length > 0 &&
+      bid.params.stream.protocols &&
+      bid.params.stream.protocols.length > 0) : true));
```

Comment:
This section readability should be improved.

Reply:
Added an initial comment and improved some instructions.